data/blog/battery-life.mdx

```diff
@@ -0,0 +1,86 @@
+- Control plane: communication with the [coordination server](https://tailscale.com/blog/how-tailscale-works/#the-control-plane-key-exchange-and-coordination)
+- DNS: both UDP and DNS-over-HTTPS
+- Port-mapping: probes done as part of [NAT traversal](https://tailscale.com/blog/how-nat-traversal-works/)
+- Logging: diagnostics and metrics uploaded by the client
+
+This instrumentation allows us to do two things:
+
+1. [Derive the cellular radio power state](https://github.com/tailscale/tailscale/pull/7777): while it’s a very simple model (assuming a 3G radio and that Tailscale is the only thing generating traffic), it allows us to have a single number that we can aggregate easily.
+2. [Logging](https://github.com/tailscale/tailscale/pull/7498): we can keep track of the type of network traffic done by a client with high-precision. These logs are somewhat expensive to maintain and upload, thus it’s only done for a subset of users. They also very detailed, thus building visualization tooling is key.
+
+With this instrumentation in place, we can have a [workflow](https://en.wikipedia.org/wiki/OODA_loop) for tackling battery life investigations:
+
+1. Collect high-level aggregated statistics (e.g. the fraction of time spent in high-power states), broken down by Tailscale version
+2. If a new version or cohort of clients shows a significant change, pick representative clients and look at their network usage.
+3. Based on the observations, make changes to the clients (or possibly change the instrumentation).
+4. Deploy changes, and see if they have the desired effect.
+
+All of this instrumentation has landed relatively recently, but we’ve already been able to use it to investigate [some](https://github.com/tailscale/tailscale/issues/7617) [issues](https://github.com/tailscale/tailscale/issues/7850):
+
+<figure className="float-none text-center">
+  <Image
+    src="/images/battery-life/sockstats-tx.png"
+    width="1021"
+    height="370"
+    alt="Graph showing WireGuard handshake attempts every 5 seconds."
+  />
+  <figcaption>
+    _If you keep sending packets every 5 seconds, you’re going to have a bad time_
+  </figcaption>
+</figure>
+
+Tech industry blog posts stereotypically end with “We rewrote our app in language X/ported it to framework Y/adopted methodology Z and all our problems were solved!”. It would be nice if all this work would lead us to a single root cause that would let us say “Tailscale’s battery life is negligible now.” However, based on our experience thus far, the much more likely outcome is that there will be many small issues or occasional interactions between sub-systems that lead to excessive battery use. It’s also likely that we’ll need to improve our tooling as we discover new things. The hope is that we’ve now established the infrastructure to be able to investigate such issues more easily, as well as keep an eye out for any regressions.
```

Comment:
```suggestion
Tech industry blog posts stereotypically end with “We rewrote our app in language X/ported it to framework Y/adopted methodology Z and all our problems were solved!”. Unfortunately, we're not there. It would be nice if all this work would lead us to a single root cause that would let us say “Tailscale’s battery life is negligible now.” Based on our experience thus far, the much more likely outcome is that there will be many small issues or occasional interactions between sub-systems that lead to excessive battery use. It’s also likely that we’ll need to improve our tooling as we discover new things.

We’ve now established the infrastructure to be able to investigate battery issues more easily, as well as keep an eye out for any regressions. The hope now is that we'll be able to see noticeable improvements as we tackle these issues.
```

Reply:
"Unfortunately, we're not there." isn't quite the right sentiment -- I think such absolutist conclusions are wrong/marketing fluff. Tweaked the text a bit.